### ISSUE TITLE
Probable bug in initializer list use in SIMD inline functions

### ISSUE THREAD

**Issue by valassi** [madgraph5/madgraph4gpu#339]
I realised there are probably one or more bugs in the way I used initializer lists for SIMD compiler vector extensions, eg
https://github.com/madgraph5/madgraph4gpu/blob/5ac285cebc04cd7eb4d6679c6a3a5e2645803d30/epochX/cudacpp/ee_mumu/src/mgOnGpuVectors.h#L220

Essentially I have coded assuming that "fptype_v{x}" would initialize all elements of the vector to x, but actually only the first element is x and the others are zero.

This needs to be debugged. It is also the reason why I added a new testmisc test infrastructure in #336

**Comment by valassi** [comment]
Maybe the cause of #338, maybe not

**Comment by valassi** [comment]
This is now fixed by PR #364, which I just merged. I am closing this issue.

NOTE: this patch fixes some bugs, but the bugs were actually harmless so far. The bug is essentially that only the first element of an array was initialized to a given value, with the subsequent ones initialized to 0. However, this function was essentially only used with the given value equal to 0, so eveything was correctly initialized to 0 anyway. Waht I mean is that no change in average ME was observed or is expected to be observed.

**Comment by roiser** [comment]
Hi @valassi , thanks for pointing me to this closed issue, as we were discussing there are a few places in the code where eg. a SIMD type is initialised with ```{0}```, this is not a problem when the initialising value is 0 as by default all values of the SIMD array will be 0, just the code above sets only the first array value. 

With my current work I stumbled over the case where I needed to initialise all entries of a SIMD type with non 0 and then the above will not work as only the first entry will be the initialised value and all other set to 0. The proposed solution in this case is to use ```std::fill_n``` e.g. 

```std::fill_n(&mysimd[0], nelems, value);```

the disadvantage here is that it can't be constexpr anymore. So maybe one should only use it for non 0 init values and for 0 just do an empty ```{}``` but the latter is a minor issue as also ```{0}``` is doing the right thing anyway. Its maybe just a bit confusing.

**Comment by valassi** [comment]
Hi @roiser, this old issue was fixed with this simple approach in PR #364
```
inline
cxtype_v cxmake( const fptype& r, const fptype_v& i )
{
  //return cxtype_v( fptype_v{r}, i ); // THIS WAS A BUG! #339
  return cxtype_v( fptype_v{0} + r, i ); // IIII=0000+r=rrrr
}
```
Check the changes in that PR there are also many other tests and examples

I would use this approach (which is probably constexpr) and avoid an inflation of different notations. I would avoid `std::fill` in particular. And certainly I would not change now code which has worked for two years.

**Comment by roiser** [comment]
I didn't know that a scalar type can be added like this to a SIMD one, that looks good, thanks. Reading a bit about it e.g. in 
[here](https://www.open-std.org/jtc1/sc22/wg21/docs/papers/2014/n4184.pdf), an implicit conversion happens but only in this case as the SIMD type is left hand. 
Still for readability I would remove the 0 in e.g. ```fptype_v{0}``` because this is syntactically wrong and misleading. But as said above as long as its a 0 it will be the same as ```fptype_v{}```.
